crates/ruff_python_formatter/tests/snapshots/format@statement__class_definition.py.snap

```diff
@@ -513,7 +513,30 @@ class QuerySet(AltersData):
  
  
  class Test(
-@@ -159,20 +158,17 @@
+@@ -94,7 +93,6 @@
+ 
+ 
+ class Test:
+-
+     """Docstring"""
+ 
+ 
+@@ -111,14 +109,12 @@
+ 
+ 
+ class Test:
+-
+     """Docstring"""
+ 
+     x = 1
+ 
+ 
+ class Test:
+-
```

Comment:
Can we add a few tests where we have comments before the docstring (including new or empty lines)

Reply:
I've tested it out locally with different combinations of comments, newline and docstrings. It seems that the Black preview style is a bit inconsistent with the stable style.

<details><summary>Original source code:</summary>
<p>

```python
class OneBlankLineDocstring:

    """This is a docstring."""


class DocstringWithComment0:
    # This is a comment
    """This is a docstring."""


class DocstringWithComment1:
    # This is a comment

    """This is a docstring."""


class DocstringWithComment2:

    # This is a comment
    """This is a docstring."""


class DocstringWithComment3:

    # This is a comment

    """This is a docstring."""


class DocstringWithComment4:


    # This is a comment


    """This is a docstring."""
```

</p>
</details> 

<details><summary>Black stable formatting:</summary>
<p>

```python
class NormalDocstring:

    """This is a docstring."""


class DocstringWithComment0:
    # This is a comment
    """This is a docstring."""


class DocstringWithComment1:
    # This is a comment

    """This is a docstring."""


class DocstringWithComment2:
    # This is a comment
    """This is a docstring."""


class DocstringWithComment3:
    # This is a comment

    """This is a docstring."""


class DocstringWithComment4:
    # This is a comment

    """This is a docstring."""
```

</p>
</details> 

<details><summary>Black preview formatting:</summary>
<p>

```python
class OneBlankLineDocstring:
    """This is a docstring."""


class DocstringWithComment0:
    # This is a comment
    """This is a docstring."""


class DocstringWithComment1:
    # This is a comment

    """This is a docstring."""


class DocstringWithComment2:

    # This is a comment
    """This is a docstring."""


class DocstringWithComment3:

    # This is a comment

    """This is a docstring."""


class DocstringWithComment4:

    # This is a comment

    """This is a docstring."""
```

</p>
</details>

What I'm noticing is that the rules change if there's a comment in between the class header and docstring in a way that for `n1` and `n2` number of blank lines before a comment and docstring respectively the following holds true:
1. If `n1` / `n2` is 0, keep it 0
2. If `n1` / `n2` is 1, keep it 1
3. If `n1` / `n2` > 1, then reduce it to 1

It's probably another preview style which is keeping the newline before the comment (`allow_empty_first_line_before_block_or_comment`).